### DISCUSSION TITLE
Efficiently load sub-relationships for already fetched objects.

### DISCUSSION THREAD

**Discussion by fschulze** [sqlalchemy/sqlalchemy#9090] [Usage Questions] [answered]
First some pseudo-code:

```
class Container:
   items = relationship(Item)

class Item:
   subitems = relationship(SubItem)

class SubItem:
   pass
```

Due to the way our application is structured, I have an instance of ``Container`` for which the ``items`` relationship to ``Item`` is eagerly loaded before the place where I know that I need the objects from the sub-relationship ``Item.subitems``. Now I need to efficiently load ``subitems`` for all already loaded objects.

I tried to recreate the query with:
```
session.query(Item)
.join(Container.items)
.filter(Container.id == self.id)
.options(orm.subqueryload(Item.subitems))
```
But that loads the objects from the already fetched ``items`` again, where I only want to load ``subitems`` for each of the existing objects.

Is there a way to do what subqueryload (or any other loading strategy) does for an already loaded relationship?

**Comment by zzzeek** [comment, marked as answer]
if you run that query and are not using populate_existing, it won't actually "load" any Item objects, the rows will come in and as their primary key matches Item objects already present in the Session, they will be skipped.  The unloaded Item.subitems relationships will however be loaded.

if you want to reduce the columns coming over the wire for Item you could do:

```py
session.query(Item)
.join(Container.items)
.filter(Container.id == self.id)
.options(orm.load_only(Item.id), orm.subqueryload(Item.subitems))
```

**Reply by zzzeek** [reply]
also if Container -> Item is  Container.id -> Item.container_id, you can drop the join for Container:

```py
session.query(Item)
.filter(Item.container_id == self.id)
.options(orm.load_only(Item.id), orm.subqueryload(Item.subitems))
```

**Reply by CaselIT** [reply]
also consider replacing `subqueryload` for `selectinload`. Rationale https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#subquery-eager-loading
